src/lib/integration_test_lib/malleable_error.ml

```diff
@@ -0,0 +1,345 @@
+open Core_kernel
+open Async_kernel
+
+(* module Test_error = Test_error *)
+
+(** The is a monad which is conceptually similar to `Deferred.Or_error.t`,
+ *  except that there are 2 types of errors which can be returned at each bind
+ *  point in a computation: soft errors, and hard errors. Soft errors do not
+ *  effect the control flow of the monad, and are instead accumulated for later
+ *  extraction. Hard errors effect the control flow of the monad in the same
+ *  way an `Error` constructor for `Or_error.t` would.
+ It remains similar to Deferred.Or_error.t in that it is a specialization of Deferred.Result.t
+ *)
+
+module Hard_fail = struct
+  type t =
+    { hard_error: Test_error.internal_error
+    ; soft_errors: Test_error.internal_error list }
+  [@@deriving eq, sexp_of, compare]
+end
+
+module Accumulator = struct
+  type 'a t =
+    {computation_result: 'a; soft_errors: Test_error.internal_error list}
+  [@@deriving eq, sexp_of, compare]
+end
+
+type 'a t = ('a Accumulator.t, Hard_fail.t) Deferred.Result.t
+
+module T = Monad.Make (struct
+  type nonrec 'a t = 'a t
+
+  let return a =
+    Deferred.return (Ok {Accumulator.computation_result= a; soft_errors= []})
+
+  (* 
+
+('a, 'b) result Deferred.t ->
+f:('c -> ('d, 'e) result Deferred.t) -> 
+('f, 'b) result Deferred.t 
+
+*)
+  let bind res ~f =
+    match%bind res with
+    | Ok {Accumulator.computation_result= prev_result; soft_errors= err_list}
+      -> (
+        let execProg = f prev_result in
+        match%bind execProg with
```

Comment:
You can just write `match%bind f prev_result with ...` here to avoid an extra value binding.

Reply:
Also, you should use `match%map ... with ...` here since you are returning in all branches of the match. It's minor, but doing this actually reduces the number of async scheduler cycles for each round of `Malleable_error.bind` application (because, in async, every `Deferred.bind` application stages the next part of the computation for the next scheduler cycle, and returning like this will introduce an extra under the hood).